### PR TITLE
Fix: arrow-parens & destructuring/default params (fixes #3353)

### LINKED ISSUE #3353: arrow-parens: as-needed doesn't account for destructuring arguments
```js
let x = ({item}) => item
let y = ([item]) => item
```

Both of these require parentheses, but `arrow-parens: as-needed` reports them as not needing parens.

BabelJS repl tests: [object destructuring], and [array destructuring].

[object destructuring]: http://babeljs.io/repl/#?experimental=true&evaluate=false&loose=false&spec=false&code=let%20x%20%3D%20%7Bitem%7D%20%3D%3E%20item%0A

[array destructuring]: http://babeljs.io/repl/#?experimental=true&evaluate=false&loose=false&spec=false&code=let%20y%20%3D%20%5Bitem%5D%20%3D%3E%20item%0A

### PR DESCRIPTION
**Pull request by xjamundx**

### CODE REVIEW

File: tests/lib/rules/arrow-parens.js [hunk 1]
```diff
@@ -29,6 +29,9 @@ var valid = [
     { code: "() => {}", options: ["as-needed"], ecmaFeatures: { arrowFunctions: true } },
     { code: "a => {}", options: ["as-needed"], ecmaFeatures: { arrowFunctions: true } },
     { code: "a => a", options: ["as-needed"], ecmaFeatures: { arrowFunctions: true } },
+    { code: "([a, b]) => {}", options: ["as-needed"], ecmaFeatures: { arrowFunctions: true, destructuring: true } },
+    { code: "({ a, b }) => {}", options: ["as-needed"], ecmaFeatures: { arrowFunctions: true, destructuring: true } },
+    { code: "(a = 10) => {}", options: ["as-needed"], ecmaFeatures: { arrowFunctions: true, destructuring: true, defaultParams: true } },
```
Comment (xjamundx) [inline tests/lib/rules/arrow-parens.js]: destructuring, defaultParams, anything else that we might need to test for this case?

Comment (michaelficarra) [inline tests/lib/rules/arrow-parens.js]: Rest parameter.

**Comment by xjamundx** [comment]
What kind of parents are we? arrow-parents.....fixing

File: tests/lib/rules/arrow-parens.js (hunk 1 above)
Comment (xjamundx) [inline tests/lib/rules/arrow-parens.js]: cool!

**Comment by xjamundx** [comment]
Should be good now?

**Comment by ilyavolodin** [comment]
LGTM

**Pull request merged**